sklearn/metrics/_ranking.py

```diff
@@ -1703,15 +1703,20 @@ def top_k_accuracy_score(
     y_type = type_of_target(y_true, input_name="y_true")
     if y_type == "binary" and labels is not None and len(labels) > 2:
         y_type = "multiclass"
-    y_score = check_array(y_score, ensure_2d=False)
-    y_score = column_or_1d(y_score) if y_type == "binary" else y_score
-    check_consistent_length(y_true, y_score, sample_weight)
-
     if y_type not in {"binary", "multiclass"}:
         raise ValueError(
             f"y type must be 'binary' or 'multiclass', got '{y_type}' instead."
         )
+    y_score = check_array(y_score, ensure_2d=False)
+    if y_type == "binary":
+        if y_score.ndim == 2 and y_score.shape[1] != 1:
+            raise ValueError(
+                "`y_true` is binary while y_score is 2d with more than one class. If"
+                " `y_true` does not contain all the labels, `labels` must be provided."
```

Comment:
Let's be more informative:

```suggestion
                f"`y_true` is binary while y_score is 2d with {y_score.shape[1]}"
                "classes. If `y_true` does not contain all the labels, `labels`"
                " must be provided."
```

I accepted, the test will need to be updated accordingly.